backend/src/books/books.service.ts

```diff
@@ -18,4 +65,30 @@ export class BooksService {
       isbn: doc.isbn,
     }));
   }
+
+  async uploadBooks(file: { buffer: Buffer }): Promise<BookUploadResult> {
+    const stream = bufferToStream(file.buffer);
+    return new Promise(async (resolve, reject) => {
+      Papa.parse(stream, {
+        header: true,
+        skipEmptyLines: 'greedy',
+        complete: async result => {
+          try {
+            const books = result.data.map(book => new BookDto(book));
+            const { books: validatedBooks, errors } = await validateBooks(
+              books,
+            );
+            await this.booksRepository.save(validatedBooks, { chunk: 1000 });
+            resolve({
+              booksToUpload: books.length,
+              booksUploaded: validatedBooks.length,
+              errors,
+            });
+          } catch (e) {
+            reject(e);
```

Comment:
I would suggest using Nestjs's Logger to log errors => `Logger.error()` , it can be imported from the `@nestjs/common` module